R/chat-async.R

```diff
@@ -0,0 +1,291 @@
+#' chat <- new_chat_async_openai()
+#' chat$register_tool(
+#'   fun = rnorm,
+#'   name = "rnorm",
+#'   description = "Drawn numbers from a random normal distribution",
+#'   arguments = list(
+#'     n = tool_arg(
+#'       type = "integer",
+#'       description = "The number of observations. Must be a positive integer."
+#'     ),
+#'     mean = tool_arg(
+#'       type = "number",
+#'       description = "The mean value of the distribution."
+#'     ),
+#'     sd = tool_arg(
+#'       type = "number",
+#'       description = "The standard deviation of the distribution. Must be a non-negative number."
+#'     )
+#'   )
+#' )
+#' chat$chat("
+#'   Give me five numbers from a random normal distribution.
+#'   Briefly explain your work.
+#' ") %...>% cat()
+new_chat_async_openai <- function(system_prompt = NULL,
+                                  base_url = "https://api.openai.com/v1",
+                                  api_key = openai_key(),
+                                  model = "gpt-4o-mini",
+                                  quiet = TRUE) {
+  check_string(system_prompt, allow_null = TRUE)
+  check_string(base_url)
+  check_string(api_key)
+  check_string(model)
+  check_bool(quiet)
+
+  system_prompt <- system_prompt %||%
+    "You are a helpful assistant from New Zealand who is an experienced R programmer"
+
+  ChatAsyncOpenAI$new(
+    base_url = base_url,
+    model = model,
+    api_key = api_key,
+    system_prompt = system_prompt,
+    quiet = quiet
+  )
+}
+
+#' @rdname new_chat_async_openai
+ChatAsyncOpenAI <- R6::R6Class("ChatOpenAI",
+  public = list(
+    #' @param system_prompt A system prompt to set the behavior of the assistant.
+    #' @param base_url The base URL to the endpoint; the default uses ChatGPT.
+    #' @param api_key The API key to use for authentication. You generally should
+    #'   not supply this directly, but instead set the `OPENAI_API_KEY` environment
+    #'   variable.
+    #' @param model The model to use for the chat; defaults to GPT-4o mini.
+    #' @param quiet If `TRUE` does not print output as its received.
+    initialize = function(base_url, model, api_key, system_prompt, quiet = TRUE) {
+      private$base_url <- base_url
+      private$model <- model
+      private$api_key <- api_key
+      private$quiet <- quiet
+
+      private$add_message(list(
+        role = "system",
+        content = system_prompt
+      ))
+    },
+
+    #' @description Submit text to the chatbot, and receive a promise that
+    #'   resolves with the response all at once. If not in quiet mode, the
+    #'   response will be printed to stdout as it is received.
+    #' @param text The text to send to the chatbot.
+    #' @param quiet Whether to emit the response to stdout as it is received. If
+    #'   `NULL`, then the value of `quiet` set when the chat object was created
+    #'   will be used.
+    #' @returns A promise that resolves to a string (probably Markdown).
+    chat = function(text, quiet = NULL) {
```

Comment:
Could this be a `chat_async()` method on the existing class?

Or I wonder if we should more separate what you the methods that you might want to call for interactive experience vs the methods you might call if you're wrapping in another tool? OTOH shiny is like to be the primary use case for async, so we could just design the minimal interface for it?